File: 프로그래머스/3/43163. 단어 변환/단어 변환.swift
import Foundation

var ans = Int.max
var visited = [Bool]() // 방문 체크

func solution(_ begin:String, _ target:String, _ words:[String]) -> Int {
    visited = Array(repeating: false, count: words.count)
    dfs(begin, target, words, 0)
    
    return ans == Int.max ? 0 : ans // 변환할 수 있으면 ans, 변환할 수 없으면 0
}
    
func dfs(_ begin: String, _ target: String, _ words: [String], _ cnt: Int) {
    if begin == target {
        // ans를 작은 값으로 갱신
        ans = ans > cnt ? cnt : ans
        return
    }
    else {
        // begin을 기준으로 모든 단어를 방문
        for i in 0..<words.count {
            if visited[i] {
                continue
            }

            let beginArr = Array(begin).map{String($0)}
            let wordCache = Array(words[i]).map{String($0)}
            var numOfDifferent = 0

            // begin과 다음 단어 간 틀린 알파벳 개수 비교
            for j in 0..<beginArr.count {
                if beginArr[j] != wordCache[j] {
                    numOfDifferent += 1
                }
            }
            
            if numOfDifferent == 1 {
                visited[i] = words[i] == target ? false : true
                dfs(words[i], target, words, cnt + 1)
            }
        }
    }
}

File: 프로그래머스/3/43163. 단어 변환/단어 변환.py
# 이해
# begin에서 출발해서 target으로 도착하기 까지 최소 단계를 return
# 이동은 words의 단어로 변환하는 것만 가능하다. 한 글자만 다른 경우 이동이 가능하다.
# 변환할 수 없는 경우 return 0

# 풀이
# 최소 단계이나 DFS도 가능할 듯하다.
# visited를 생성한다.
# DFS
    # 탈출 조건은 target이 달성된 경우.
        # 이동 횟수를 최소값으로 갱신한다.
    # for word in words
    # 현재 word를 방문하지 않았다면,
        # 현재 word를 방문 처리
        # dfs(cnt + 1)

INF = int(1e9)
answer = INF

def solution(begin, target, words):
    visited = [False] * len(words)
    dfs(begin, target, words, visited, 0)
    return 0 if answer == INF else answer

def dfs(begin, target, words, visited, cnt):
    global answer
    if begin == target:
        answer = min(answer, cnt)
    
    for idx, word in enumerate(words):
        if not visited[idx] and check(begin, word):
            visited[idx] = True
            dfs(word, target, words, visited, cnt + 1)
            visited[idx] = False

def check(a_word, b_word):
    tmp = 0
    for i in range(len(a_word)):
        if a_word[i] != b_word[i]:
            tmp += 1
    
    return True if tmp == 1 else False
    
    
    
    
    
    
    
    
    